doc/development/source/orange-demo/orangedemo/OWDataSamplerA.py

```diff
@@ -3,16 +3,20 @@
 import numpy
 
 import Orange.data
-from Orange.widgets import widget, gui
+from Orange.widgets.widget import OWWidget, Input, Output
+from Orange.widgets import gui
 
-class OWDataSamplerA(widget.OWWidget):
+class OWDataSamplerA(OWWidget):
     name = "Data Sampler"
     description = "Randomly selects a subset of instances from the data set"
     icon = "icons/DataSamplerA.svg"
     priority = 10
 
-    inputs = [("Data", Orange.data.Table, "set_data")]
-    outputs = [("Sampled Data", Orange.data.Table)]
+    class Inputs:
+        data = Input("Data", Orange.data.Table)
+
+    class Outputs:
+        sample = Output("Sampled Data", Orange.data.Table)
```

Comment:
.. and then write `self.Settings.variable` instead of `self.variable` everywhere in the widget? I would really hate that.

Having classes like `Error` or `Inputs` requires some magic, which I hate (although it's fun to write it). The reason for having them was to use identifiers instead of literal constants. The `Settings` class would have entirely different purpose, it would be about making the programmer aware that `self.variable` is saved as a setting. I don't see so much value in this; surely not enough to warrant another piece of magic.

As for properties... Generic example is

```
class P:
    def __init__(self,x):
        self.__x = x

    def get_x(self):
        return self.__x

    def set_x(self, x):
        self.__x = x

    x = property(get_x, set_x)
```

We don't want to write `set_x` and `get_x`. However, when the property is created (e.g. with `x = property(...)`) the name to which it is bound is not yet knownm, so we can't create the setter and getter. Metaclasses to the rescue again.

```
class Setting(property):
    def __init__(self, default, fget=None, fset=None):
        super().__init__(fget, fset)
        self.default = default

class WidgetMeta(type):
    def __new__(mcs, name, bases, kwargs):
        cls = super().__new__(mcs, name, bases, kwargs)

        for name, setting in cls.__dict__.items():
            if isinstance(setting, Setting):
                uname = "_" + name
                default = setting.default  # needed for closure; setting.default wouldn't work in getter
                def setter(widget, value):
                    setattr(widget, uname, value)
                def getter(widget):
                    return getattr(widget, uname, default)
                # fget and fset are read-only attributes of property, hence we must
                # create a new Setting
                setattr(cls, name, Setting(default, getter, setter))

        return cls

class Widget(metaclass=WidgetMeta):
    x = Setting(42)
    y = Setting("foo")

w = Widget()
print(w.x)
w.x = 50
print(w.x)
```

This is ugly and it doesn't have any benefits over the existing `Setting` - it still involves metamagic and PyCharm still doesn't know the types of `x` and `y`.